server/cluster_worker.go

```diff
@@ -0,0 +1,414 @@
+	resp, err := c.s.client.Txn(context.TODO()).
+		If(c.s.leaderCmp()).
+		Then(clientv3.OpDelete(jobKey)).
+		Commit()
+	if err != nil {
+		return errors.Trace(err)
+	} else if !resp.Succeeded {
+		return errors.Errorf("pop first job failed")
+	}
+	return nil
+}
+
+func (c *raftCluster) updateJobStatus(job *pd_jobpd.Job, status pd_jobpd.JobStatus) error {
+	jobKey := makeJobKey(c.clusterRoot, job.GetJobId())
+	job.Status = status.Enum()
+	jobValue, err := proto.Marshal(job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	resp, err := c.s.client.Txn(context.TODO()).
+		If(c.s.leaderCmp()).
+		Then(clientv3.OpPut(jobKey, string(jobValue))).
+		Commit()
+	if err != nil {
+		return errors.Trace(err)
+	} else if !resp.Succeeded {
+		return errors.Errorf("pop first job failed")
+	}
+	return nil
+}
+
+func (c *raftCluster) handleJob(job *pd_jobpd.Job) error {
+	log.Debugf("begin to handle job %v", job)
+
+	// TODO: if the job status is running, check this job whether
+	// finished or not in raft server.
+	if job.GetStatus() == pd_jobpd.JobStatus_Pending {
+		if err := c.updateJobStatus(job, pd_jobpd.JobStatus_Running); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	req := job.GetRequest()
+	switch req.AdminRequest.GetCmdType() {
+	case raft_cmdpb.AdminCommandType_ChangePeer:
+		return c.handleChangePeer(job)
+	default:
+		log.Errorf("invalid job command %v, ignore", req)
+		return nil
+	}
+}
+
+func (c *raftCluster) chooseStore(bestStores []metapb.Store, matchStores []metapb.Store) metapb.Store {
+	var store metapb.Store
+	// Select the store randomly, later we will do more better choice.
+
+	if len(bestStores) > 0 {
+		store = bestStores[rand.Intn(len(bestStores))]
+	} else {
+		store = matchStores[rand.Intn(len(matchStores))]
+	}
+
+	return store
+}
+
+func (c *raftCluster) handleAddPeerReq(region *metapb.Region) (*metapb.Peer, error) {
+	peerID, err := c.s.idAlloc.Alloc()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		// The best stores are that the region has not in.
+		bestStores []metapb.Store
+		// The match stores are that region has not in these stores
+		// but in the same node.
+		matchStores []metapb.Store
+	)
+
+	mu := &c.mu
+	mu.RLock()
+	defer mu.RUnlock()
+
+	// Find a proper store which the region has not in.
+	for _, store := range mu.stores {
+		storeID := store.GetStoreId()
+		nodeID := store.GetNodeId()
+
+		existNode := false
+		existStore := false
+		for _, peer := range region.Peers {
+			if peer.GetStoreId() == storeID {
+				// we can't add peer in the same store.
+				existStore = true
+				break
+			} else if peer.GetNodeId() == nodeID {
+				existNode = true
+			}
+		}
+
+		if existStore {
+			continue
+		} else if existNode {
+			matchStores = append(matchStores, store)
+		} else {
+			bestStores = append(bestStores, store)
+		}
+	}
+
+	if len(bestStores) == 0 && len(matchStores) == 0 {
+		return nil, errors.Errorf("find no store to add peer for region %v", region)
+	}
```

Comment:
no need, and the error message contains the region info, if we move to `chooseStore`, we should pass region as another arg too. 

and later we will improve the choosing strategy, so maybe later.